File: packages/python/agentkit/keys.py
"""Ed25519 over canonical bytes.

Keys are handled as raw 32-byte values, the same way the kernel stores them, and wrapped
into whatever encoding the crypto library wants only at the moment it is needed.
"""

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .canonical import canonical_bytes

RAW_KEY_LENGTH = 32


def _raw(value, label: str) -> bytes:
    if isinstance(value, (bytes, bytearray)):
        return bytes(value)
    if isinstance(value, str):
        return bytes.fromhex(value)
    raise TypeError(f"{label} must be bytes or a hex string")


def generate_key_pair() -> tuple[bytes, bytes]:
    """Returns (public_key, private_key) as raw 32-byte values."""
    private = Ed25519PrivateKey.generate()
    from cryptography.hazmat.primitives import serialization

    return (
        private.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        ),
        private.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        ),
    )


def sign_payload(private_key, payload) -> bytes:
    raw = _raw(private_key, "private key")
    if len(raw) != RAW_KEY_LENGTH:
        raise ValueError(f"ed25519 private key must be {RAW_KEY_LENGTH} bytes, received {len(raw)}")
    return Ed25519PrivateKey.from_private_bytes(raw).sign(canonical_bytes(payload))


def verify_payload(public_key, payload, signature) -> bool:
    try:
        raw = _raw(public_key, "public key")
        sig = _raw(signature, "signature")
        Ed25519PublicKey.from_public_bytes(raw).verify(sig, canonical_bytes(payload))
        return True
    except (InvalidSignature, ValueError, TypeError):
        return False

File: packages/merchant/lib/keys.js
"use strict";

const {
    createPrivateKey,
    createPublicKey,
    generateKeyPairSync,
    sign: nodeSign,
    verify: nodeVerify,
} = require("node:crypto");

const { canonicalBytes } = require("./canonical");

/**
 * Ed25519 over canonical bytes.
 *
 * Keys are handled as raw 32-byte values and wrapped into DER only at the moment
 * node:crypto needs them, so what you store is the key rather than some library's
 * encoding of it. The kernel stores them the same way; the two must match.
 */

const RAW_PUBLIC_KEY_LENGTH = 32;
const RAW_PRIVATE_KEY_LENGTH = 32;

const DER_PUBLIC_PREFIX = Buffer.from("302a300506032b6570032100", "hex");
const DER_PRIVATE_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");

function toBuffer(value, label) {
    if (Buffer.isBuffer(value)) return value;
    if (typeof value === "string") return Buffer.from(value, "hex");
    throw new TypeError(`${label} must be a Buffer or a hex string`);
}

/** A fresh agent identity. Persist both halves: the id a mandate names is derived from it. */
function generateKeyPair() {
    const { publicKey, privateKey } = generateKeyPairSync("ed25519");
    const spki = publicKey.export({ type: "spki", format: "der" });
    const pkcs8 = privateKey.export({ type: "pkcs8", format: "der" });
    return {
        publicKey: Buffer.from(spki.subarray(spki.length - RAW_PUBLIC_KEY_LENGTH)),
        privateKey: Buffer.from(pkcs8.subarray(pkcs8.length - RAW_PRIVATE_KEY_LENGTH)),
    };
}

function toPrivateKeyObject(raw) {
    if (raw.length !== RAW_PRIVATE_KEY_LENGTH) {
        throw new RangeError(
            `ed25519 private key must be ${RAW_PRIVATE_KEY_LENGTH} bytes, received ${raw.length}`,
        );
    }
    return createPrivateKey({
        key: Buffer.concat([DER_PRIVATE_PREFIX, raw]),
        format: "der",
        type: "pkcs8",
    });
}

function toPublicKeyObject(raw) {
    if (raw.length !== RAW_PUBLIC_KEY_LENGTH) {
        throw new RangeError(
            `ed25519 public key must be ${RAW_PUBLIC_KEY_LENGTH} bytes, received ${raw.length}`,
        );
    }
    return createPublicKey({
        key: Buffer.concat([DER_PUBLIC_PREFIX, raw]),
        format: "der",
        type: "spki",
    });
}

function signPayload(privateKey, payload) {
    const key = toPrivateKeyObject(toBuffer(privateKey, "private key"));
    return nodeSign(null, canonicalBytes(payload), key);
}

function verifyPayload(publicKey, payload, signature) {
    try {
        const key = toPublicKeyObject(toBuffer(publicKey, "public key"));
        return nodeVerify(null, canonicalBytes(payload), key, toBuffer(signature, "signature"));
    } catch {
        return false;
    }
}

module.exports = { generateKeyPair, signPayload, verifyPayload };
